### ISSUE TITLE
a11y issues 

### ISSUE THREAD

**Issue by JuliannaR** [cds-snc/ircc-rescheduler#161]
Here is a quick list of issues related to the reschedule project. Please note that this is not an exhaustive list. (time spent: 20 minutes/SME + 10 minutes/user testing) 

Meth: Tested with WAVE, Nu Validator, Colour Contrast Analyzer, keyboard and user tested for adaptive technology (screen reader and voice rec).

1. Contrast Minimum:  Buttons fail (Start, Continue, Review) - Can be testing with colour contrast analyzer.

![cdsreschcolour](https://user-images.githubusercontent.com/35537308/41546727-69911ee8-72ec-11e8-97c6-47846830dfc2.JPG)

2. Info and Relationships: Headings
        a. https://rescheduler.cds-snc.ca/ Request a new Canadian Citizenship appointment should be an H1. 
        b. please do not use hidden headings. Re: First verify your identity and tell us why you need a new  
            appointment. 

3. Language of Page: Language declaration is not set on pages. 

4. Error Identification: If you select only 1 or 2 dates in the calendar there error description is not sufficient to identify the issue. I would suggest indicating that 3 date selections are required and indicate that it is required.

     "Sorry, there was a problem with the information you submitted.
      Please take another look at the:

      Calendar"

5. Info and Relationships/Labels and Instructions: Remove date button is not associated with which date is being removed. This causes a burden and makes it difficult for a user to understand which dates they are removing or adding. 

6. Link Purpose: https://rescheduler.cds-snc.ca/review and https://rescheduler.cds-snc.ca/calendar 
   Adjacent links go to the same URL. change, change, change, and review, review, review

![cdsreschadjacentlinks](https://user-images.githubusercontent.com/35537308/41546687-50d5b9e0-72ec-11e8-9890-2df2a264270c.JPG)

   When adjacent links go to the same location (such as a linked 
   product image and an adjacent linked product name that go to the same product page) this results in 
   additional navigation and repetition for keyboard and screen reader users.

Usability Findings:

Chrome & JAWS 2018
Trying to correct a date seems to crash the calendar. You need to start over completely when clicking Remove date.

IE and JAWS 2018 
July seemed fine until I clicked the remove button on one of the dates. It erased ALL of them.

## Issues

- [x] Contrast Minimum: Buttons fail
- [ ] Info and Relationships: Headings
- [x] Language of Page
- [x] Error Identification
- [x] Info and Relationships/Labels and Instructions
- [x] Link Purpose

**Comment by kyliehavelock** [comment]
This is super helpful @JuliannaR , thank you so much for taking the time! 🙏 

Most of these issues were also picked up in our recent accessibility audit, which you can read here for reference: https://github.com/cds-snc/ircc-rescheduler/issues/162
Today we processed these outcomes as a team and created a bunch of new development stories to resolve the highlighted issues, which we'll be working through in the next few weeks.

Regarding your usability findings on JAWS with Chrome and IE, the team have tried to replicate these but we haven't been able to. These weren't picked up in our accessibility audit either. Are you able to share any more detail so we can visualise the problems?

**Comment by JuliannaR** [comment]
I can look into it. They were found by an adaptive tech user in our program who uses JAWS in order to work. I'll go over the issues with him and see if I can provide more insight on that front. 

Glad to see progress on this and consistent findings. :) Keep working hard, being inclusive is the best thing we can do for Canadians.

**Comment by pcraig3** [comment]
Hi @JuliannaR, thanks for doing this for us! As @kyliehavelock mentioned, we did a recent accessibility audit that picked up a lot of these issues (but not all of them!).

I'll create a bullet list here so that we can see the progress we're making as we address these one at a time. (I've left off the JAWS findings for now because I don't know how to reproduce or investigate those at present.)

## Issues

- [ ] Contrast Minimum: Buttons fail
- [ ] Info and Relationships: Headings
- [ ] Language of Page
- [ ] Error Identification
- [ ] Info and Relationships/Labels and Instructions
- [ ] Link Purpose

I'll update the issue as we make fixes.

**Comment by nschonni** [comment]
Error messages also fail colour contrast https://webaim.org/resources/contrastchecker/?fcolor=E8112D&bcolor=F5F5F5

**Comment by nschonni** [comment]
Axe plug-in also flags a bunch of these https://axe-core.org/

**Comment by nschonni** [comment]
https://marcysutton.github.io/a11y-and-ci might be a good reference too

**Comment by JuliannaR** [comment]
Glad to see the progress on this! Thanks for the feedback and keeping us in the loop.

**Comment by nschonni** [comment]
Error messages look good now, but your "Beta" badge fails contrast

**Comment by nschonni** [comment]
Sorry, I try to write it up properly when I'm home, but I noticed some heading issues on the Calendar page a few days ago if you haven't already dealt with them.

**Comment by pcraig3** [comment]
Yeah, I think I know what you mean. There's an axe warning that the calendar month + year name isn't being properly marked up with the correct heading level.

**Comment by nschonni** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}  Yup, and the selected dates section jumps to an H3                                                                                                                                                                        Mistyped on my CrackBerry                                                                                                                                                                           From: notifications@github.comSent: July 17, 2018 6:15 PMTo: ircc-rescheduler@noreply.github.comReply to: reply@reply.github.comCc: nschonni@gmail.com; comment@noreply.github.comSubject: Re: [cds-snc/ircc-rescheduler] a11y issues (#161)  Yeah, I think I know what you mean. There's an axe warning that the calendar month + year name isn't being properly marked up with the correct heading level.

**Comment by pcraig3** [comment]
Hey, @nschonni. A few more fixes in today.

- beta badge colour updated: https://github.com/cds-snc/ircc-rescheduler/pull/274
- heading level added to calendar months: https://github.com/cds-snc/ircc-rescheduler/pull/271

# Link Purpose

Fixed since https://github.com/cds-snc/ircc-rescheduler/pull/265

- Announces the name of the heading to be modified
- Links to the correct section on the page based on the link that was clicked
- Works in English and French

**Comment by pcraig3** [comment]
# Request a new Canadian Citizenship appointment should be an H1

Fixed since #295.

# please do not use hidden headings.

Hi @JuliannaR, do you mind expanding on this, as I was under the impression this was okay.

The problem we’re trying to solve is that [on the registration page](https://rescheduler-dev.cds-snc.ca/register), there’s nothing that could really be used as an H1 — by which I mean “a sentence that tells you why you are on this page” — and in our usability testing we've found we don't need one. Visual users understand our interface pretty well and almost always make it through this page without much trouble. (Especially since those error messages improved :D) 

However, for screenreader users, I was under the impression that they would scan for headers, and that the H1 on the page would orient them: both in terms of preceding the rest of the page's content, and in  announcing the meaning of the page.

In this case, we have a hidden heading on the form page “first provide some basic information”, which visual users don’t need but may help orient those with screenreaders. 

Is there a reason this is a bad idea / an anti-pattern? 

I’m by no means an expert screenreader user and I’ve also never watched someone who relies on a screenreader go through our site, so I’m just not sure.

**Comment by pcraig3** [comment]
Going to close this issue as we are halting active development on this project.
If the hidden headings are an ongoing issue, I would suggest opening new issue.
